### DISCUSSION TITLE
Views V2

### DISCUSSION THREAD

**Discussion by mjashanks** [Budibase/budibase#2999] [Ideas]
[Views](https://docs.budibase.com/docs/views) have been with us for a long time. However, since we added the "Define Filters" option to `Data Providers`, they have lost some of their purpose.

However, views are still the only way that we can achieve “aggregation” functions on internal tables

Please give your opinions below on views.

### How we use views today

**Aggregation functions**
- `sum` of a column.
- `count` of records
- `average` of a column

**Filters**
We can also apply a filter to a dataset (without applying an aggregation). This has largely been superseded by “Define Filters”. However, this is still needed for performing aggregates on the filtered dataset.

**Group by**
Allows to apply aggregates per column.

**Charts**
Whilst "Charts" is not a feature of views per say, this is one of the primary uses of a view - especially with the combination of a group by. 

For example, you might want to display a bar chart, where each bar is a Group.

We should consider how views and charts interact.

## Limitations

**Applying a dynamic filter**
For example, we cannot apply a filter like
`Creation Date > {{ fromDate }}` - where `fromDate` is a parameter passed into the view.
https://github.com/Budibase/budibase/discussions/2986

**Group by relationship**
If a `Project` has many `Tasks`, we cannot apply a `Count of Tasks per Project`
https://github.com/Budibase/budibase/discussions/2763
https://github.com/Budibase/budibase/discussions/3171

**Group by, then by**
This would allow us to perform 2 or more groupings in one query.
https://github.com/Budibase/budibase/discussions/3426

**Group by Formula**
https://github.com/Budibase/budibase/discussions/3678

**"Having"**
"Having" (SQL terminology) is the ability to apply a filter on an aggregate function. For example:

`Tasks, Grouped By Project, Having Count > 5`  =  show me projects that have more than 5 tasks.

**Formulas**
Views do not support formula fields. 
https://github.com/Budibase/budibase/discussions/1033
https://github.com/Budibase/budibase/discussions/2271
https://github.com/Budibase/budibase/discussions/3170
Group by formula is particularly useful, for creating on-the fly groups. e.g. "Group by Month" (where date is a timestamp). This was raised by Sypher and here: https://github.com/Budibase/budibase/discussions/3678

**Filters do not handle null**
https://github.com/Budibase/budibase/issues/2042

**Sorting and pagination**
This is irrelavant if views are only to be used for calculations. However, we are unable to sort data by any given column and then paginate.
https://github.com/Budibase/budibase/discussions/2235

**Column projection**
Again, irrelevant if using calculations only. But, we cannot choose which columns a view returns: https://github.com/Budibase/budibase/discussions/1299.

This was raised in https://github.com/Budibase/budibase/issues/3268, with a really good point - restricting access to certain columns.

**Comment by daawaan4x** [comment]
Hello, are there any updates regarding this?

**Comment by xnohat** [comment]
Hi, currently at version 1.0.98 still cannot filter by formula column in View ?

**Reply by marco-santulli** [reply]
hi mike,

I think it would also be very useful to allow views for the external databases, like postgres or mysql.  what do you think?

M

**Reply by daawaan4x** [reply]
@marco-santulli

> hi mike,
> 
> I think it would also be very useful to allow views for the external databases, like postgres or mysql.  what do you think?
> 
> M

BudibaseDB Views are just the same as custom queries for MySQL and PostgreSQL

**Comment by hmmhmmhm** [comment]
![스크린샷 2022-04-03 오후 1 32 25](https://user-images.githubusercontent.com/38493722/161411622-843fc361-7dd9-4854-8ea6-312b776d40eb.png)
i need this

**Comment by marco-santulli** [comment]
@daawaan sorry replying here because github is playing up with replies...

I thought views on the internal DB supported paginations?  
Also, a view would be picked up by budibase automatically, without the need to write and maintain a query no?

M

**Comment by Zero2XIII** [comment]
Any progress on this point, or is there any known workaround?:

> Group by, then by
>This would allow us to perform 2 or more groupings in one query.
>https://github.com/Budibase/budibase/discussions/3426

**Comment by Brrrradley** [comment]
Any progress on this point as well? Can't proceed without a workaround.

> Formulas
Views do not support formula fields.
https://github.com/Budibase/budibase/discussions/1033
https://github.com/Budibase/budibase/discussions/2271
https://github.com/Budibase/budibase/discussions/3170
Group by formula is particularly useful, for creating on-the fly groups. e.g. "Group by Month" (where date is a timestamp). This was raised by Sypher and here: https://github.com/Budibase/budibase/discussions/3678

**Comment by hurrifan1** [comment]
As far as #1299 and #3268 and the need to limit columns in a view, what is the recommended way of handling that in the absence of that feature? Just make two linked tables where one has the "public" columns and the other has the "restricted" columns?

**Comment by phamalda** [comment]
Hi, @mjashanks. Looks like your initial [Views](https://docs.budibase.com/building-apps/data/views) link is broken. I think that should be [Views](https://docs.budibase.com/docs/views)?

**Reply by melohagan** [reply]
Thanks! I've updated the link

**Comment by wflamme** [comment]
Look at Lotus Notes/Domino's response docs and categorized views.. We probably could tweak this using couchDb standard reduce views/grouping and custom components?
<img width="1180" height="189" alt="budibase - tweak group view" src="https://github.com/user-attachments/assets/af706042-c56f-47cd-a6bb-c2e37cc9d8ba" />
<img width="1179" height="183" alt="budibase - tweak group view2" src="https://github.com/user-attachments/assets/83983a78-d535-45e9-bbcc-93dbacb3335a" />
